Hardware/HDD/AbstractStorage.cs

```diff
@@ -0,0 +1,254 @@
+    private Sensor sensorDiskWriteRate { get; set; } = null;
+
+    protected AbstractStorage(StorageInfo info, string name, string firmwareRevision, string id, int index, ISettings settings)
+      : base(name, new Identifier(id, index.ToString(CultureInfo.InvariantCulture)), settings) {
+
+      this.storageInfo = info;
+      this.firmwareRevision = firmwareRevision;
+      this.index = index;
+      this.count = 0;
+
+      string[] logicalDrives = WindowsStorage.GetLogicalDrives(index);
+      List<DriveInfo> driveInfoList = new List<DriveInfo>(logicalDrives.Length);
+      foreach (string logicalDrive in logicalDrives) {
+        try {
+          DriveInfo di = new DriveInfo(logicalDrive);
+          if (di.TotalSize > 0)
+            driveInfoList.Add(new DriveInfo(logicalDrive));
+        }
+        catch (ArgumentException) {
+        }
+        catch (IOException) {
+        }
+        catch (UnauthorizedAccessException) {
+        }
+      }
+      driveInfos = driveInfoList.ToArray();
+    }
+
+    public static AbstractStorage CreateInstance(string deviceId, uint driveNumber, ulong diskSize, int scsiPort, ISettings settings) {
+      StorageInfo info = WindowsStorage.GetStorageInfo(deviceId, driveNumber);
+      info.DiskSize = diskSize;
+      info.DeviceId = deviceId;
+      info.Scsi = string.Format(@"\\.\SCSI{0}:", scsiPort);
+
+      if (info == null || info.Removable || info.BusType == Interop.StorageBusType.BusTypeVirtual || info.BusType == Interop.StorageBusType.BusTypeFileBackedVirtual)
+        return null;
+
+      if (info.BusType == Interop.StorageBusType.BusTypeAta || info.BusType == Interop.StorageBusType.BusTypeSata)
+        return ATAStorage.CreateInstance(info, settings);
+
+      if (info.BusType == Interop.StorageBusType.BusTypeNvme)
+        return NVMeGeneric.CreateInstance(info, settings);
+
+      return StorageGeneric.CreateInstance(info, settings);
+    }
+
+    protected virtual void CreateSensors() {
+      if (driveInfos.Length > 0) {
+        usageSensor = new Sensor("Used Space", 0, SensorType.Load, this, settings);
+        ActivateSensor(usageSensor);
+      }
+
+      sensorDiskWriteActivity = new Sensor("Write Activity", 32, SensorType.Load, this, settings);
+      sensorDiskTotalActivity = new Sensor("Total Activity", 33, SensorType.Load, this, settings);
+      ActivateSensor(sensorDiskWriteActivity);
+      ActivateSensor(sensorDiskTotalActivity);
+
+      sensorDiskReadRate = new Sensor("Read Rate", 34, SensorType.Throughput, this, settings);
+      sensorDiskWriteRate = new Sensor("Write Rate", 35, SensorType.Throughput, this, settings);
+      ActivateSensor(sensorDiskReadRate);
+      ActivateSensor(sensorDiskWriteRate);
+    }
+
+    public override HardwareType HardwareType {
+      get { return HardwareType.HDD; }
+    }
+
+    protected abstract void UpdateSensors();
+
+
+    public class PerformanceValue {
+      public ulong perfValue { get; set; } = 0;
```

Comment:
These properties should be Capitalized and these names are meaningless. What do they do exactly and what are the differences?